datasets/fr/amf_regulatory_sanctions/crawler.py

```diff
@@ -0,0 +1,161 @@
+from typing import Any, Generator, Optional, Tuple, Union
+
+from normality import collapse_spaces
+from zavod import Context, helpers as h
+import re
+from html import unescape
+from urllib.parse import urljoin
+from lxml import html
+
+
+CLEAN_ENTITY = re.compile(r"(<br />\r\n| et |;)", re.IGNORECASE)
+BASE_URL = "https://www.amf-france.org"
+
+
+def parse_json(context: Context) -> Generator[dict, None, None]:
+    response = context.fetch_json(context.data_url)
+    if "data" not in response:
+        context.log.info("No data available.")
+        return
+
+    context.log.info(
+        f"Fetched {len(response['data'])} results."
+    )  # Log the number of results
+    for item in response["data"]:
+        yield item
+
+
+def get_value(
+    data: dict, keys: Tuple[str, ...], default: Optional[Any] = None
+) -> Union[Optional[Any], Any]:
+    for key in keys:
+        val = data.pop(key, None)
+        if val is not None:
+            return val
+    return default
+
+
+PREFIXES = [
+    "Sociétés",
+    "Société",
+    "société",
+    "Banque",
+    "la société",
+    "Caisse",
+    "cabinet",
+    "La sociétés",
+    "les CABINETS",
+    "les sociétés",
+    "LA",
+    "de gestion",
+    "cabinets",
+    "SCS",  # Société en commandite simple
+    r"M\.?",
+    "Mme",
+    r"MM\.?",
+    "Madame",
+    "Monsieur",
+    "Monsieur",
+    "Melle",
+]
+
+
+comma_prefix_anon = "(,|et) +(" + "|".join(PREFIXES) + r")[\. \b]+[A-Z]\b"
+prefix_anon = "(" + "|".join(PREFIXES) + r")[\. \b]+[A-Z]\b"
+comma_anon = r"(,|et) +[A-Z]\b"
+
+
+def clean_names_str(names_str: str) -> Optional[str]:
+    names_str = re.sub(comma_prefix_anon, "", names_str)
+    names_str = re.sub(prefix_anon, "", names_str)
+    names_str = re.sub(comma_anon, "", names_str)
+    names_str = collapse_spaces(names_str)
+    return names_str
+
+
+def process_entity(
+    context: Context,
+    title: str,
+    listing_date: str,
+    name: str,
+    theme: str,
+    link: Optional[str],
+    item: dict,
+) -> None:
+    """Process a legal entity entry."""
+    entity = context.make("LegalEntity")
+    entity.id = context.make_id(name, listing_date)
```

Comment:
when I want to be fairly sure I'm generating the same ID in different parts of the code, I like to factor that out to a function with a clear name and call it from both places

```suggestion
def entity_id(context, name, listing_date):
    return context.make_id(name, listing_date)


def process_entity(
    context: Context,
    title: str,
    listing_date: str,
    name: str,
    theme: str,
    link: Optional[str],
    item: dict,
) -> None:
    """Process a legal entity entry."""
    entity = context.make("LegalEntity")
    entity.id = entity_id(context, name, listing_date)
```